Dateï¼š20251029  
Languageï¼š[EN](#Fantastic-Genomic-Biomarkers-and-Where-to-Find-Them-Practical-Course-part-IIX) / [ä¸­æ–‡](#ç”Ÿç‰©æ¨™è¨˜ç‰©èˆ‡å®ƒå€‘çš„ç”¢åœ°å¯¦ä½œèª²ç¨‹å…«)

## Fantastic Genomic Biomarkers and Where to Find Them Practical Course (part IIX)

## Main Content of the Course

#### Using `hap.py` to compare the analysis results of HaplotypeCaller and Mutect2 to understand their performance and accuracy in differnt scenarios.

> [!Warning]
> ### Prerequisite: Prepare the results of the 3 WGS data from the previous tutorial, each processed with two variant calling methods: HaplotypeCaller and Mutect2!! 
> #### If not yet completed, please first copy the TA's files.
> ```markdown=
> cd /work/username
> mkdir vcf_for_happy
> cd vcf_for_happy
> rsync -avzP /work/evelyn92/result/tutorial_result/tutorial_VC/S14_VC/SRR13076392_S14_L002.sorted.markdup.m2.vcf.gz ./
> rsync -avzP /work/evelyn92/result/tutorial_result/tutorial_VC/S15_VC/SRR13076396_S15_L002.sorted.markdup.m2.vcf.gz ./
> rsync -avzP /work/evelyn92/result/tutorial_result/tutorial_VC/S18_VC/SRR13076396_S18_L002.sorted.markdup.m2.vcf.gz ./
> rsync -avzP /work/evelyn92/result/tutorial_result/tutorial_VC/S14_VC/SRR13076392_S14_L002.sorted.markdup.hc.vcf.gz ./
> rsync -avzP /work/evelyn92/result/tutorial_result/tutorial_VC/S15_VC/SRR13076393_S15_L002.sorted.markdup.hc.vcf.gz ./
> rsync -avzP /work/evelyn92/result/tutorial_result/tutorial_VC/S18_VC/SRR13076396_S18_L002.sorted.markdup.hc.vcf.gz ./
> ```

> [!Important]
> #### What is `hap.py`?
> 
> `Hap.py` (Haplotype Comparison Tools) is a tool used for comparing genomic variants. It is often utilized to evaluate the accuracy of variant calling algorithms, especially in identifying variants such as SNPs and Indels in somatic or germline cells. `hap.py` can be used to compare variant calling results with known standards (e.g., a gold standard VCF) to assess the sensitivity, precision, and other metrics of variant detection methods.

### Step 1 : Run `hap.py`
[Shell script](https://github.com/Jacob-s-Lab/2025-Biomarkers/blob/main/hap.sh)
1. Copy the executable files and reference files required for the class.
```markdown=
cd /work/username
rsync -avz /work/evelyn92/hap.sh ./
rsync -avz /work/evelyn92/KnownPositives_hg38_Liftover.vcf ./
rsync -avz /work/evelyn92/High-Confidence_Regions_v1.2.bed.gz ./
``` 

2. Open `hap.sh` and enter the correct path names.  \
(1) ğŸ”´ **Red** part: Change the TA's username to your own.  \
(2) ğŸ”µ **Blue** file names: Modify according to your files.
(3) ğŸŸ¡ **Yellow** part: Change to your own file path.

![è¢å¹•æ“·å–ç•«é¢ 2025-10-15 152422](https://hackmd.io/_uploads/SypGyRhaex.png)
![è¢å¹•æ“·å–ç•«é¢ 2025-10-15 153242](https://hackmd.io/_uploads/SJ6710hpxe.png)


> [!Important]
> #### Reminder: 
> **The bcftools command is used to remove multiallelic variants from the VCF file so that `hap.py` can read it.**

3. Run `hap.sh`.
```
sbatch hap.sh
```

4. Get the results: a total of 11 files.
- output_prefix.runinfo.json
- output_prefix.metrics.json.gz
- output_prefix.roc.Locations.SNP.csv.gz
- output_prefix.roc.Locations.SNP.PASS.csv.gz
- output_prefix.roc.Locations.INDEL.csv.gz 
- output_prefix.roc.Locations.INDEL.PASS.csv.gz
- output_prefix.roc.all.csv.gz
- output_prefix.summary.csv
- output_prefix.extended.csv
- output_prefix.vcf.gz
- output_prefix.vcf.gz.tbi

### Step 2: Use `rocplot.sh` results to create an ROC plot.
[Shell script](https://github.com/Jacob-s-Lab/2025-Biomarkers/blob/main/rocplot.sh)

1. Copy the executable files required for the class.
```markdown=
rsync -avz /work/evelyn92/rocplot/rocplot.sh ./
rsync -avz /work/evelyn92/rocplot/rocplot_test.Rscript ./
```
2. Enter R and load the required packages.
```markdown=
R
install.packages("ggplot2")
61
install.packages("tools")
q()
n
```

3. Change to the correct path.
(1) ğŸ”µ **Blue** part: Make sure whether it is a HaplotypeCaller or Mutect2 file.
(2) ğŸ”´ **Red** part: Replace with your own username.
![è¢å¹•æ“·å–ç•«é¢ 2025-10-15 215643](https://hackmd.io/_uploads/rJljF_76agl.png)

5. Run `rocplot.sh`.
```
sbatch rocplot.sh
```
5. The results are saved in the rocplot_HC or rocplot_M2 folders, each containing two plots.
For example:

- Comparison of **SNP results** for S14, S15, S18 using Mutect2:
![hap_plot.SNP](https://hackmd.io/_uploads/BJBePXa6ex.png)


- Comparison of **Indel results** for S14, S15, S18 using Mutect2:
![hap_plot.INDEL](https://hackmd.io/_uploads/rkm-v76Txl.png)



6. You can also open IGV to compare the results from `hap.py`.
The results are saved in ```output_prefix.vcf.gz.```
- Comparison between tools:
Example: Comparison of SNPs from S18 Mutect2 and HaplotypeCaller. At position 241,884,674, the variant can be found with HaplotypeCaller, but not with Mutect2.
![è¢å¹•æ“·å–ç•«é¢ 2025-10-15 214642](https://hackmd.io/_uploads/Hy6D8XpTeg.png)

- Comparison between samples: 
Example: Comparison of SNPs for S14, S15, S18 using Mutect2.
![æ¨£æœ¬é–“æ¯”è¼ƒ](https://hackmd.io/_uploads/SkmUH7pTeg.png)






--------------------------------
# ç”Ÿç‰©æ¨™è¨˜ç‰©èˆ‡å®ƒå€‘çš„ç”¢åœ°å¯¦ä½œèª²ç¨‹(å…«)
## æœ¬æ¬¡èª²ç¨‹ä¸»è¦å…§å®¹
    
#### å­¸ç¿’åˆ©ç”¨`hap.py`æ¯”è¼ƒ Haplotypecaller å’Œ Mutect2 çš„åˆ†æçµæœï¼Œä»¥äº†è§£å®ƒå€‘åœ¨ä¸åŒæƒ…å¢ƒä¸‹çš„æ€§èƒ½å’Œæº–ç¢ºæ€§ã€‚


> [!Warning]
> ### å‰æƒ…æè¦ï¼šéœ€å…ˆæº–å‚™å¥½ä¹‹å‰ tutorial åšå®Œçš„ä¸‰å€‹ WGS data åˆ†åˆ¥åšå…©ç¨® variant calling: HaplotypeCaller åŠ Mutect2 çš„çµæœï¼ï¼
> #### å¦‚æœå°šæœªå®Œæˆè«‹å…ˆè¤‡è£½åŠ©æ•™çš„æª”æ¡ˆï½
>  ```markdown=
> cd /work/username
> mkdir vcf_for_happy
> cd vcf_for_happy
> rsync -avzP /work/evelyn92/result/tutorial_result/tutorial_VC/S14_VC/SRR13076392_S14_L002.sorted.markdup.m2.vcf.gz ./
> rsync -avzP /work/evelyn92/result/tutorial_result/tutorial_VC/S15_VC/SRR13076396_S15_L002.sorted.markdup.m2.vcf.gz ./
> rsync -avzP /work/evelyn92/result/tutorial_result/tutorial_VC/S18_VC/SRR13076396_S18_L002.sorted.markdup.m2.vcf.gz ./
> rsync -avzP /work/evelyn92/result/tutorial_result/tutorial_VC/S14_VC/SRR13076392_S14_L002.sorted.markdup.hc.vcf.gz ./
> rsync -avzP /work/evelyn92/result/tutorial_result/tutorial_VC/S15_VC/SRR13076393_S15_L002.sorted.markdup.hc.vcf.gz ./
> rsync -avzP /work/evelyn92/result/tutorial_result/tutorial_VC/S18_VC/SRR13076396_S18_L002.sorted.markdup.hc.vcf.gz ./
> ```

> [!Important]
> #### ä»€éº¼æ˜¯`hap.py`ï¼Ÿ
> `hap.py`ï¼ˆHaplotype Comparison Toolsï¼‰æ˜¯ä¸€å€‹ç”¨æ–¼æ¯”è¼ƒåŸºå› çµ„è®Šç•°çš„å·¥å…·ã€‚å®ƒç¶“å¸¸è¢«ç”¨ä¾†è©•ä¼° variant calling ç®—æ³•çš„æº–ç¢ºæ€§ï¼Œç‰¹åˆ¥æ˜¯åœ¨é«”ç´°èƒæˆ–ç”Ÿæ®–ç´°èƒä¸­çš„ SNPs å’Œ Indels ç­‰è®Šç•°çš„è­˜åˆ¥ã€‚`hap.py` å¯ä»¥ç”¨ä¾†å°æ¯” variant calling çµæœèˆ‡å·²çŸ¥çš„æ¨™æº–ç­”æ¡ˆï¼ˆä¾‹å¦‚ gold standard VCF ï¼‰ï¼Œä»¥è©•ä¼°è®Šç•°æª¢æ¸¬æ–¹æ³•çš„éˆæ•åº¦ (Recall)ã€ç²¾ç¢ºåº¦ (Precision)ç­‰æŒ‡æ¨™ã€‚

### step 1 : åŸ·è¡Œ`hap.py`
[Shell script](https://github.com/Jacob-s-Lab/2025-Biomarkers/blob/main/hap.sh)

1. è¤‡è£½ä¸Šèª²æ‰€éœ€åŸ·è¡Œæª”èˆ‡æ¨™æº–æª”ã€‚
```markdown=
cd /work/username
rsync -avz /work/evelyn92/hap.sh ./
rsync -avz /work/evelyn92/KnownPositives_hg38_Liftover.vcf ./
rsync -avz /work/evelyn92/High-Confidence_Regions_v1.2.bed.gz ./
```

2. æ‰“é–‹`hap.sh`ä¸¦å¯«å…¥æ­£ç¢ºçš„è·¯å¾‘åç¨±ã€‚
   (1)ğŸ”´ **ç´…è‰²** éƒ¨åˆ†ï¼šåŠ©æ•™çš„ **username** æ”¹æˆè‡ªå·±çš„ã€‚
   (2)ğŸ”µ **è—è‰²** çš„æª”æ¡ˆåç¨±ï¼šéš¨ä½ çš„æª”æ¡ˆæ”¹è®Šã€‚
   (3)ğŸŸ¡ **é»ƒè‰²** éƒ¨åˆ†ï¼šæ”¹æˆè‡ªå·±çš„æª”æ¡ˆè·¯å¾‘ã€‚
![è¢å¹•æ“·å–ç•«é¢ 2025-10-15 152422](https://hackmd.io/_uploads/SypGyRhaex.png)
![è¢å¹•æ“·å–ç•«é¢ 2025-10-15 153242](https://hackmd.io/_uploads/SJ6710hpxe.png)

> [!Important]
> #### æé†’ï¼š
> **bcftools å‘½ä»¤çš„éƒ¨åˆ†æ˜¯å°‡ vcf æª”ä¸­çš„ multialelic å»é™¤ï¼Œä»¥ä¾¿`hap.py`å¯ä»¥è®€å–ã€‚**
> 

3. åŸ·è¡Œ```hap.sh```
```
sbatch hap.sh
```
4. å¾—åˆ°çµæœ : å…±11å€‹æª”æ¡ˆã€‚
- output_prefix.runinfo.json
- output_prefix.metrics.json.gz
- output_prefix.roc.Locations.SNP.csv.gz
- output_prefix.roc.Locations.SNP.PASS.csv.gz
- output_prefix.roc.Locations.INDEL.csv.gz 
- output_prefix.roc.Locations.INDEL.PASS.csv.gz
- output_prefix.roc.all.csv.gz
- output_prefix.summary.csv
- output_prefix.extended.csv
- output_prefix.vcf.gz
- output_prefix.vcf.gz.tbi




### step 2 : åˆ©ç”¨`rocplot.sh`çµæœè£½ä½œ ROC åœ–
[Shell script](https://github.com/Jacob-s-Lab/2025-Biomarkers/blob/main/rocplot.sh)

1. è¤‡è£½ä¸Šèª²æ‰€éœ€åŸ·è¡Œæª”ã€‚
```markdown=
rsync -avz /work/evelyn92/rocplot/rocplot.sh ./
rsync -avz /work/evelyn92/rocplot/rocplot_test.Rscript ./
```
2. é€²å…¥Rï¼Œè¼‰å…¥éœ€è¦çš„ package
```markdown=
R
install.packages("ggplot2")
61
install.packages("tools")
q()
n
```

3. æ›´æ”¹æ­£ç¢ºè·¯å¾‘
   (1)ğŸ”µ **è—è‰²** çš„éƒ¨åˆ†ï¼šé ˆæ³¨æ„æ˜¯ **Haplotypecaller æˆ–æ˜¯ Mutect2 çš„æª”æ¡ˆ**ã€‚
   (2)ğŸ”´ **ç´…è‰²** çš„éƒ¨åˆ†ï¼šéœ€æ›æˆè‡ªå·±çš„ **username**ã€‚
![è¢å¹•æ“·å–ç•«é¢ 2025-10-15 215643](https://hackmd.io/_uploads/rJljF_76agl.png)

5. åŸ·è¡Œ`rocplot.sh`
```
sbatch rocplot.sh
```
5. å¾—åˆ°çµæœå­˜æ–¼ rocplot_HC æˆ– rocplot_M2 è³‡æ–™å¤¾ï¼Œè£¡é¢å„æœ‰å…©å¼µåœ–ã€‚


ä¾‹å¦‚ï¼š
- S14, S15, S18 Mutect2 åœ¨ SNP çš„çµæœæ¯”è¼ƒï¼š
![hap_plot.SNP](https://hackmd.io/_uploads/SJz1Ym6Tee.png)




- S14, S15, S18 Mutect2 åœ¨ Indel çš„çµæœæ¯”è¼ƒï¼š
![hap_plot.INDEL](https://hackmd.io/_uploads/H17xYQTaxx.png)


6. ä¹Ÿå¯ä»¥æ‰“é–‹ IGV æ¯”è¼ƒ`hap.py`çš„çµæœã€‚
çµæœå„²å­˜æ–¼```output_prefix.vcf.gz```
- å·¥å…·é–“çš„æ¯”è¼ƒï¼š
    èˆ‰ä¾‹ï¼šS18 Mutect2 åŠ Haplotypecaller çš„ SNP æ¯”è¼ƒï¼Œåœ¨ chr1 çš„241,884,674é€™å€‹ä½é»å¯ä»¥åœ¨ Haplotypecaller è¢«æ‰¾åˆ°ï¼Œä½†åœ¨ Mutect2 æ‰¾ä¸åˆ°ã€‚
![è¢å¹•æ“·å–ç•«é¢ 2025-10-15 214642](https://hackmd.io/_uploads/S1CWFQpTee.png)
- æ¨£æœ¬é–“çš„æ¯”è¼ƒï¼š
    èˆ‰ä¾‹ï¼šS14, S15, S18 Mutect2 çš„ SNP æ¯”è¼ƒã€‚
![æ¨£æœ¬é–“æ¯”è¼ƒ](https://hackmd.io/_uploads/BkJmFQapee.png)
